packages/components/legacy_text/src/legacy_text.tsx

```diff
@@ -12,107 +19,78 @@ export type LegacyTextVariant =
   | "bodySmallRegularItalic"
   | "caption"
   | "error"
-  | "headline1"
-  | "headline2"
-  | "headline3"
-  | "input"
-  | "label"
-  | "overline"
-  | "subtitle";
+  | "input";
+
+export type LegacyTextVariant =
+  | BodyVariant
+  | HeadingVariant
+  | SubheadingVariant;
 
 interface OwnProps {
   variant?: LegacyTextVariant;
 }
 
-export type TextProps = BoxProps & TypographyProps & OwnProps;
+type HeadingVariantProps = { variant: HeadingVariant } & Omit<HeadingProps, 'size'>;
+type SubheadingVariantProps = { variant: SubheadingVariant } & Omit<SubheadingProps, 'size'>;
+type BodyVariantProps = { variant: BodyVariant } & Omit<BodyProps, 'size'>;
+
+export type TextProps = (HeadingVariantProps | SubheadingVariantProps | BodyVariantProps) & OwnProps;
 
-export const LegacyText: React.FC<TextProps> = ({
-  children,
-  variant,
-  ...props
-}): React.ReactElement => {
```

Comment:
Had to remove spreading of `props` in the component arguments to allow type discrimination to work correctly here